dd-sdk-android-ndk/src/main/cpp/utils/signal-monitor.c

```diff
@@ -45,98 +63,74 @@ struct signal {
-                                &global_previous_sigaction[i]);
-        if (success != 0) {
-            return false;
-        }
-    }
-    return true;
-}
-
-bool try_to_install_handlers(JNIEnv *env) {
-    if (global_env != NULL) {
-        return true; // already installed
-    }
-    global_env = env;
-
-    if (!configure_signal_stack()) {
-        return false;
-    }
+        if (signal == signum) {
+            struct sigaction previous = global_previous_sigaction[i];
+            // From sigaction(2):
+            //  > If act is non-zero, it specifies an action (SIG_DFL, SIG_IGN, or a
+            //  handler routine)
+            if (previous.sa_flags & SA_SIGINFO) {
+                // This handler can handle signal number, info, and user context
+                // (POSIX). From sigaction(2): > If this bit is set, the handler
+                // function is assumed to be pointed to by the sa_sigaction member of
+                // struct sigaction and should match the proto- type shown above or as
+                // below in EXAMPLES.  This bit should not be set when assigning SIG_DFL
+                // or SIG_IGN.
+                previous.sa_sigaction(signum, info, user_context);
+            } else if (previous.sa_handler == SIG_DFL) {
+                // raise to trigger the default handler. It cannot be called directly.
+                raise(signum);
 
-    if (!configure_global_sigaction()) {
-        return false;
+            } else if (previous.sa_handler != SIG_IGN) {
+                // This handler can only handle to signal number (ANSI C)
+                void (*previous_handler)(int) = previous.sa_handler;
+                previous_handler(signum);
+            }
+        }
     }
 
-    return override_native_signal_handlers();
-}
-
-
-bool install_signal_handlers(JNIEnv *env) {
-    pthread_mutex_lock(&handlerMutex);
-    bool installed = try_to_install_handlers(env);
-    pthread_mutex_unlock(&handlerMutex);
-
-    return installed;
+    pthread_mutex_unlock(&mutex);
 }
 
-
-void uninstall_signal_handlers() {
-    if (global_env == NULL)
-        return;
-
-    pthread_mutex_lock(&handlerMutex);
-
-    if (global_env == NULL) {
-        pthread_mutex_unlock(&handlerMutex);
-        return;
-    }
+void handle_signal(int signum, siginfo_t *info, void *user_context) {
     const size_t signals_array_size = handled_signals_size();
     for (int i = 0; i < signals_array_size; i++) {
         const int signal = handled_signals[i].signal_value;
-        sigaction(signal, &global_previous_sigaction[i], 0);
+        if (signal == signum) {
+            crash_signal_intercepted(signal,
+                                     handled_signals[i].signal_name,
+                                     handled_signals[i].signal_error_message);
+            break;
+        }
     }
-    global_env = NULL;
 
-    pthread_mutex_unlock(&handlerMutex);
+    uninstall_signal_handlers();
```

Comment:
Honestly...that's the way it was (found it) :). This file had several issues that I corrected so I guess this might be another one. If I am following the logic...what it does is to 
1. Restore the handlers to previous (original) state.
2. Simulate the call to the previous handler as it would happen if the original handlers were already there.

They should not clash but in the same time question is that if the process is already killed should we even bother with restoring them ? Or maybe this is a system internals requirement in order to be able to call the previous handler.

-